### ISSUE TITLE
Easiest way to keep styles in sync?

### ISSUE THREAD

**Issue by johnny-bit** [jade-nl/dt.styles#1]
Hi,
So far loving the styles present here and whole collection. However with changes/renames it might be a bit hard to keep them manually in sync.
What's the best (prefferably automated) way of keeping them in sync in darktable?

**Comment by jade-nl** [comment]
> However with changes/renames it might be a bit hard to keep them manually in sync.

I fully agree with that sentiment and have been thinking about that myself.

I'm still in the process of giving dt.styles its final form and up to that moment things might not be as formalised as it should be. I am working toward the following and I sure wouldn't mind your take on it:

- There is only going to be a master branch that holds 'all'. No subdividing sections in different branches
- There are going to be multiple sections (B&W, Colour, Ilford, Creative etc)
- All the files/styles will be named and described the same, no matter if they appear in multiple sections **(!!)**
- The sections will be individually downloadable. I might make combo's, not sure about that.

I have already tested and taken steps with the B&W section (locally, not pushed to GitHub yet) to achieve the above points. But it takes time to do all sections.

Not sure if you have noticed, but the Black and White section has multiple layers: *Black & White*,  *brand* and *product* This is nicely visible in the styles module in lighttable view, but in darkroom view only 2 levels are shown. I think it works, but I'm willing to be convinced otherwise.

The way darktable itself deals with styles after they are imported is somewhat limited in my opinion (ever try to remove or export multiple styles at once?). The only way to do some accounting is to install sqlitebrowser (or alike) and load data.db.

Ok here's **the TL;DR** To make sure that files will be more easily recognised I am going to make sure that the filenames, style-names and descriptions will stay the same, no matter the sections they end up in. This will take some time to do.

**Comment by jade-nl** [comment]
That wasn't as time consuming as I expected.....

Have a look and if anything is still wrong let me know!

**Comment by jade-nl** [comment]
BTW: I did not really answer the automation part of your question: 

Not sure if you are in any way familiar with sqlite3, so just ignore this if you are :-)

This might help you in removing multiple, specific entries and can also be used from within scripts (syntax probably needs to be slightly adjusted, I would use a HERE document within a bash/ksh script).

If you do want to play with this: Make sure darktable is not running and make a backup of all the files in your ~/.config/darktable directory first.

start sqlite3 with data.db:
```
$ cd ~/.config/darktable
$ sqlite3 data.db
```
this shows all entries that contain Agfa:

`sqlite> select name from styles where name like '%Agfa%' order by name;`

this shows all entries that start with Agfa:

`sqlite> select name from styles where name like 'Agfa%' order by name;
`

delete all entries that start with Agfa.
data.db holds 2 separate tables that have style data
- styles
- style_items

delete Agfa entries from style_items, needs to be done first:

`sqlite> delete from style_items where style_items.styleid in ( select id from styles where name like 'Agfa%');`

then delete the appropriate styles entries:

`sqlite> delete from styles where name like 'Agfa%';`

exit sqlite3

`sqlite> .exit`

**Comment by johnny-bit** [comment]
So if we were to keep all styles from this repo inside one dir this bash script (that can be made into oneliner) should clear all styles imported from here... Just need to make sure "our" stay out of the names here or keep them all in and re-import:

```bash
for styledir in *; do 
  if [ -d "$styledir" ]; then 
    echo "Removing $styledir styles..."; 
    sqlite3 ~/.config/darktable/data.db "delete from style_items where style_items.styleid in ( select id from styles where name like '$styledir%'); delete from styles where name like '$styledir%'"; 
  fi
done
```
I'm not good enough in LUA and darktable api, but it seems that using LUA one could also do auto-import of all styles based off of https://www.darktable.org/lua-api/index.html#darktable_styles

Do you think it's worth a try?

**Comment by jade-nl** [comment]
About the script snippet:

Something like that should work. Just tried a variation of the above and that nicely cleans all the imported stuff:

```
#!/bin/bash
STLS_SRC="/data/General/DT.Shared/Styles/dt.styles"
STLS_DB="/home/jade/.config/darktable.dev/data.db"
cd "${STLS_SRC}"
ls -d * | \
while read THIS_DIR
do
  echo "Removing ${THIS_DIR} styles..."
  sqlite3 "${STLS_DB}" "delete from style_items where style_items.styleid in ( select id from styles where name like '${THIS_DIR}%')"
  sqlite3 "${STLS_DB}" "delete from styles where name like '${THIS_DIR}%'";
done
```

About LUA:

I have looked at LUA in the past, but haven't had any good reason to use actually use it (and to be honest: I do not want to learn yet another [scripting] language). So I really don't know how easy/hard it is to use it.

If I would want to import all the style files I would probably decide to use bash and sqlite3 to do this. Haven't looked into bulk importing though (out of scope for me personally).

On my "production" side of things I only, very occasionally, use a black and white style that is not my own as a starting point. This whole dt.style project is basically for other people :-)

**Comment by johnny-bit** [comment]
I think I preffer my snippet, just need to clean it up a bit ;) Yours hardcodes locations (mine too tho, but only assumes current darktable dir). I think I'll try to learn LUA just for this (I also planned to learn lua ages ago and I'd also like to create exports... so yeah, I think I'll try lua stuff). So if I can I'll try to do some kind of universal styles clearer & later on - importer just for this ;)

**Comment by jade-nl** [comment]
The main issue has been fixed: Names and descriptions are now uniform for all styles.

@johnny-bit : It would be nice to have a platform independent importer and deleter. No pressure though :-)
